Refactor and writer

elrnv opened the pull request:
This PR is unfortunately very large, and makes a bunch of improvements (IMO) as well as adding a writer for `.obj` as well as `.mtl` files.

I am open to making changes or reintroducing some of the code/APIs that have been removed here if need be.

Below is a summary of noteworthy changes.

- Split `Obj` into a path and `ObjData`, which can be compared with `PartialEq`, since two `obj`s from different files can be equivalent.
- Updated `Mtl` to be a lazily loaded vector of actual material data. It is explicitly stored in the `ObjData` struct, and referenced by each group using Arcs instead of Cow.
- `ObjMaterial` enum wraps the Material struct. `ObjMaterial` is either a reference to a material (a name, before the materials are loaded) or an Arc pointing to the Material data stored in one of the material libs.
- Refactored the reader API to accept Read types instead of `BufRead` types, which makes the API easier to use (the user doesn't have to create `BufReader` themselves).
- The string "default" is saved as a constant and referenced by variable names `DEFAULT_OBJECT` and `DEFAULT_GROUP`.
- Added a `WriteToBuf` trait that defines the interface for structs to write their data to a buffer. This is similar to the Display trait but is specific to the Obj type. It is exposed publicly since polygons in Obj are generic.
- Added more documentation to structs and fields.
- Added a writer for the obj and mtl file formats.
- Added a round trip test with and without materials for the sponza test file.

Edit: Additional changes followed from the discussion:
- In an effort to simplify the API we removed the generic parameter for polygons throughout, replacing it with a converter from `SimplePolygon` to `genmesh::Polygon`.
- Made `load` and `save` functions accept `impl AsRef<Path>` instead of just `&Path` making them easier to use.
- Filtered the crate through clippy and rustfmt. Also added CI jobs to run rustfmt and clippy.

elrnv committed:
Add missing Error & Display impls + docs

elrnv committed:
Major refactor + add writer

- Split Obj into a path and ObjData, which can be compared with PartialEq, since two objs from different files can be equivalent.
- Updated Mtl to be a lazily loaded vector of actual material data. It is explicitly stored in the ObjData struct, and referenced by each group using Arcs instead of Cow.
- ObjMaterial enum wraps the Material struct. ObjMaterial is either a reference to a material (a name, before the materials are loaded) or an Arc pointing to the Material data stored in one of the material libs.
- Refactored the reader API to accept Read types instead of BufRead types, which makes the API easier to use (the user doesn't have to create BufReader themselves).
- The string "default" is saved as a constant and referenced by variable names DEFAULT_OBJECT and DEFAULT_GROUP.
- Added a WriteToBuf trait that defines the interface for structs to write their data to a buffer. This is similar to the Display trait but is specific to the Obj type. It is exposed publicly since polygons in Obj are generic.
- Added more documentation to structs and fields.
- Added a writer for the obj and mtl file formats.
- Added a round trip test with and without materials for the sponza test file.

elrnv committed:
Update README to indicate that writing is possible

kvark commented on src/mtl.rs:
```diff
@@ -162,26 +206,53 @@ impl<'a, I: Iterator<Item = &'a str>> Parser<I> {
             match parser.0.next() {
                 Some("newmtl") => {
-                    mtl.materials.extend(material.take());
+                    self.materials.extend(material.take().map(|m| Arc::new(m)));
```
nit: `map(Arc::new)`

kvark commented on src/mtl.rs:
```diff
@@ -279,9 +350,70 @@ impl Mtl {
+        Ok(self)
+    }
+
+    pub fn write_to_buf<W: Write>(&self, out: &mut W) -> Result<(), io::Error> {
```
nit: these days it's a tiny bit easier to write as `(&self, out: &mut impl Write)`

kvark commented on src/obj.rs:
```diff
@@ -139,15 +266,68 @@ where
+            writeln!(out, "g {}", self.name)?;
+        }
+
+        match &self.material {
```
let's not match on a reference here, just `match self.material`

kvark commented on src/obj.rs:
```diff
@@ -139,15 +266,68 @@ where
+
+/// The data model associated with each `Obj` file.
+#[derive(Debug, PartialEq)]
+pub struct ObjData<P = SimplePolygon> {
```
why do we need the default type to be set?

elrnv replied:
Solely trying to make the calls to `load` and `save` less verbose. To someone just trying to load an `Obj` it seems like unnecessary complexity to have to think about what type of polygons to use. That's just my opinion though, I am not sure about the tradeoffs here.

The default parameter here makes this call:
```rust
let obj = Obj::<SimplePolygon>::load(&Path::new("file.obj"))?;
```
into this:
```rust
let obj: Obj = Obj::load(&Path::new("file.obj"))?;
```

kvark replied:
I'm totally fine with the generic for `load` and `save` methods specifically. Does it have to be put on the whole struct? If yes, we could have an `fn load_simple` and stuff.

kvark commented on tests/round_trip.rs:
```diff
@@ -0,0 +1,53 @@
+use std::path::Path;
+
+#[test]
+fn round_trip_sponza_no_mtls() {
```
wonderful to have this!

kvark reviewed (approved):
This looks just wonderful, although I haven't tried to carefully inspect all the logic. I assume it works for some cases, and I'm happy to see the tests!
Noted a few cosmetic changes.
The biggest thing I would propose is, since you are basically rewriting the whole thing, that we refresh the formatting. We'd want to do the stable rustfmt, so we'll need to clear the config of all the things that are nightly-only. All of that, of course, needs to be in a separate commit.

Would you want to do this? Or leave for somebody to follow up?

elrnv committed:
Implement review suggestions

elrnv committed:
Add an assert checking the group merge assumption

When reading groups with multiple materials are split into unique groups
identified by an index. When writing the groups are merged once again,
and the invariant that subsequent groups with index > 0 have the same
name as their predecessors is checked.

elrnv committed:
Use `AsRef<Path>` on API calls to reduce verbosity

This commit also ensures that this change doesn't incur additional
compilation costs by keeping these generic functions small.

elrnv committed:
Run rustfmt + remove unstable features from rustfmt.toml

elrnv committed:
Address clippy recommendations

elrnv committed:
Add rustfmt and clippy to travis CI

The clippy job is allowed to fail.

elrnv committed:
Add missing exports to lib.rs + expose error struct

Missing exports include ObjMaterial and MtlLibsLoadError.
Also made the interior of MtlLibsLoadError public, since it's just a
vector of other errors.

elrnv committed:
Remove the generic Polygon parameter throughout

This makes loading and saving less verbose and simpler, at the cost of
requiring the user to explicitly call the polygon conversion code for
genmesh Polygons.

elrnv committed:
Finishing touches + cleanup

Added some missing docs, removed the unused `WriteToBuf` impl for
`Polygon<IndexTuple>`.

kvark commented on src/obj.rs:
```diff
@@ -87,22 +74,27 @@ impl WriteToBuf for Polygon<IndexTuple> {
-            3 => Ok(Polygon::PolyTri(Triangle::new(gs[0], gs[1], gs[2]))),
-            4 => Ok(Polygon::PolyQuad(Quad::new(gs[0], gs[1], gs[2], gs[3]))),
+impl SimplePolygon  {
+    fn into_genmesh_poly(self) -> Polygon<IndexTuple> {
```
could be called `into_genmesh`?

elrnv committed:
Rename into_genmesh_poly -> into_genmesh
